src/relay/pass/partial_eval.cc

```diff
@@ -0,0 +1,766 @@
+    if (const STensorNode* op = st->pstatic.as<STensorNode>()) {
+      return ConstantNode::make(op->data);
+    } else if (const STupleNode* op = st->pstatic.as<STupleNode>()) {
+      tvm::Array<Expr> fields;
+      for (const PStatic& field : op->fields) {
+        fields.push_back(Reflect(field));
+      }
+      return TupleNode::make(fields);
+    } else {
+      LOG(FATAL) << "Unknown case";
+      throw;
+    }
+  }
+
+  PStatic Reify(const Value& v, LetList* ll) const {
+    if (const TensorValueNode* op = v.as<TensorValueNode>()) {
+      return HasStatic(MkSTensor(op->data), ll->Push(ConstantNode::make(op->data)));
+    } else if (const TupleValueNode* op = v.as<TupleValueNode>()) {
+      std::vector<PStatic> fields;
+      tvm::Array<Expr> fields_dyn;
+      for (const Value& field : op->fields) {
+        PStatic ps = Reify(field, ll);
+        fields.push_back(ps);
+        fields_dyn.push_back(ps->dynamic);
+      }
+      return HasStatic(MkSTuple(fields), ll->Push(TupleNode::make(fields_dyn)));
+    } else {
+      LOG(FATAL) << "Unknown case";
+      throw;
+    }
+  }
+
+  // Constant evaluate a expression.
+  PStatic ConstEvaluate(const Expr& expr, LetList* ll) {
+    Expr infered = InferType(expr, Module(nullptr));
+    Expr fused = FuseOps(infered, 0);
+    Expr fused_infered = InferType(fused, Module(nullptr));
+    return Reify(executor_(fused_infered), ll);
+  }
+
+  Func ConstEvaluateFunc(const Expr& expr, LetList* ll) {
+    return [=](const std::vector<PStatic>& pv,
+               const Attrs& attrs,
+               const tvm::Array<Type>& type_args,
+               LetList* ll) {
+      tvm::Array<Expr> ns_args;
+      for (const PStatic& ps : pv) {
+        ns_args.push_back(ps->dynamic);
+      }
+      PStatic ns = NoStatic(CallNode::make(expr, ns_args, attrs, type_args));
+      if (StatefulOp(expr)) {
+        return ns;
+      }
+      tvm::Array<Expr> args;
+      for (const PStatic& ps : pv) {
+        if (ps->pstatic.defined()) {
+          args.push_back(Reflect(ps));
+        } else {
+          return ns;
+        }
+      }
+      return ConstEvaluate(CallNode::make(expr, args, attrs, type_args), ll);
+    };
+  }
+
+  PStatic VisitExpr_(const OpNode* op, LetList* ll) final {
+    return HasStatic(MkSFunc(ConstEvaluateFunc(GetRef<Expr>(op), ll)), GetRef<Expr>(op));
+  }
+
+  PStatic VisitExpr_(const ConstructorNode* op, LetList* ll) final {
+    Constructor c = GetRef<Constructor>(op);
+    Func f = [=](const std::vector<PStatic>& pv,
+                 const Attrs& attrs,
+                 const tvm::Array<Type>& type_args,
+                 LetList* ll) {
+      tvm::Array<Expr> dyn;
+      for (const PStatic& ps : pv) {
+        dyn.push_back(ps->dynamic);
+      }
+      return HasStatic(MkSConstructor(c, pv), ll->Push(CallNode::make(c, dyn)));
+    };
+    return HasStatic(MkSFunc(f), GetRef<Expr>(op));
+  }
+
+  PStatic VisitExpr_(const MatchNode* op, LetList* ll) final {
```

Comment:
Could you add a test case for pattern match? For example i'd like to see the following test case(really powerful):
Call `hd`  on `cons(1, cons(3, nil))`, it gets inlined to match expression, then simplified to 1.

Reply:
It doesnt support module yet. I can define my own head and call it instead of using prelude, is it fine with you? We can get module next time as it is ton of work and there is too much in this PR

Reply:
Oh, I'm fine with it. I just want to see adt + function inline + pattern match interplay together.